R/readConfig.R

```diff
@@ -5,172 +5,243 @@
-
-  # list all config files this config is based on
-  readCfg <- function(file) {
-    cfg <- read_yaml(file)
-    if (!"title" %in% names(cfg)) {
-      stop("There is no title given in the config file: ", file)
-    }
-    attr(cfg, "file") <- file
-    attr(cfg, "files") <- file
-    return(cfg)
+  # check config files that are based on the current to avoid circle dependency
+  basisOf <- attr(config, "basisOf")
+  if (any(duplicated(basisOf))) {
+    stop("There is a circle dependency in how configs are based on another: ",
+         paste(basisOf, collapse = " -> "))
   }
 
-  # Directly read the file in the given config file path - used after the final config has already been compiled
+  # Directly read the file without considering basis configs
   if (readDirect) {
     if (!file.exists(config)) {
-      stop("The config", config, "that was called to be read directly does not exist. ",
-           "If this is a restart run, your folder to restart in likely misses the config file.")
+      stop("The config", config, "that you want read directly does not exist. ",
+           "If this is a restart run, your run folder likely misses the config file.")
+    }
+    if (!is.null(configFolder)) {
+      warning("'configFolder' is ignored as the config is read directly from here: ",
+              config)
     }
-    return(readCfg(config))
+    return(.readCfg(config))
   }
 
-  configFolder <- brick.file("config")
+  # use default in config folder if available, otherwise internal default
+  internalConfigFolder <- brick.file("config")
+  if (is.null(configFolder)) {
+    configFolder <- internalConfigFolder
+  } else {
+    if (!dir.exists(configFolder)) {
+      stop("Directory of 'configFolder' doesn't exist: ", configFolder)
+    }
+  }
   defaultCfgPath <- file.path(configFolder, "default.yaml")
-
   if (!file.exists(defaultCfgPath)) {
-    stop("Default config ", defaultCfgPath, " does not exist.")
+    defaultCfgPath <- file.path(internalConfigFolder, "default.yaml")
   }
 
   # use default.yaml by default
   if (is.null(config)) {
     if (is.null(basisOf)) {
+      config <- defaultCfgPath
       message("Using default config: ", config)
     }
   }
 
   # find file to given config
+  customCfgPath <- .findCfg(config, configFolder, isFinalCfg = is.null(basisOf))
+
+  # read yaml
+  customCfg <- .readCfg(customCfgPath)
+  if (file.path(customCfgPath) == file.path(defaultCfgPath)) {
+    return(.readCfg(defaultCfgPath))
+  }
+
+  # read config on which this config is based
+  basedOn <- customCfg[["basedOn"]]
+  basedOn <- if (is.null(basedOn)) {
+    defaultCfgPath
+  } else if (length(basedOn) == 1) {
+    basedOn
+  } else {
+    stop("Don't give more than one other config that this config is based on.")
+  }
+
+  # read base config and overwrite it with given config
+  attr(basedOn, "basisOf") <- c(basisOf, customCfgPath)
+  basedOnCfg <- readConfig(basedOn, configFolder)
```

Comment:
Again it might make sense to add a comment that the chain of configs based on one another is read recursively here.

Reply:
Thanks for the hint. I extended the description of `.overwriteList` to explain the recursive function calls.